File: recon/render/compare.js
'use strict';
/*
 * The browser half of `recon.compare`, over `package.json` rather than lockfiles.
 *
 * This is a second implementation, which this repository normally treats as a
 * defect. It is allowed here on one condition, enforced by
 * `tests/test_crosscheck.py`: it must produce **byte-identical JSON** to the
 * Python `compare()` for the same inputs. A second implementation that is
 * proven to agree is a cross-check; one that is merely believed to agree is the
 * bug factory recon exists to catch. If you change either side, the test fails
 * until you change both.
 *
 * It reads manifests, not lockfiles, and that is a deliberate limit rather than
 * an accident of what is easy. A manifest is JSON in every ecosystem, while
 * pnpm and yarn lockfiles are YAML and bun's is binary — so reading manifests
 * needs no parser, no vendored dependency, and no build step, and it works for
 * repositories that commit no lockfile at all. What it cannot answer is what
 * actually got installed. Those fields come out empty rather than guessed, and
 * the page says so.
 */
(function (global) {
  var PEER = /\(.*\)$/;
  var NUMERIC = /^\d+\.\d+\.\d+.*$/;
  var NAME = /^[A-Za-z0-9._-]+$/;
  // Slashes are legal in refs (`feature/x`), so `..` is excluded by name —
  // otherwise `owner/repo@../..` walks the raw URL somewhere else.
  var REF = /^[A-Za-z0-9._/-]+$/;

  function byCodepoint(x, y) { return x < y ? -1 : x > y ? 1 : 0; }

  // `@scope/name@1.2.3` -> ['@scope/name', '1.2.3']. The peer suffix comes off
  // first or the split lands inside `(vue@3.3.4)`.
  function splitIdent(ident) {
    ident = String(ident).replace(PEER, '');
    var at = ident.lastIndexOf('@');
    if (at <= 0) return [ident, ''];
    return [ident.slice(0, at), ident.slice(at + 1)];
  }

  function aliasTarget(specifier) {
    var spec = String(specifier || '').trim();
    if (spec.indexOf('npm:') !== 0) return '';
    return splitIdent(spec.slice(4))[0];
  }

  function isPinned(specifier) {
    return Boolean(specifier) && NUMERIC.test(specifier);
  }

  function basename(name) {
    return name.charAt(0) === '@' && name.indexOf('/') > -1
      ? name.slice(name.indexOf('/') + 1) : name;
  }

  function scopeOf(name) {
    return name.charAt(0) === '@' && name.indexOf('/') > -1
      ? name.slice(0, name.indexOf('/')) : '';
  }

  function parseRepo(value) {
    var text = String(value || '').trim().replace(/\/+$/, '');
    var ref = null;
    if (text.indexOf('git@github.com:') === 0) text = text.slice('git@github.com:'.length);
    var prefixes = ['https://github.com/', 'http://github.com/', 'github.com/'];
    for (var i = 0; i < prefixes.length; i++) {
      if (text.indexOf(prefixes[i]) === 0) { text = text.slice(prefixes[i].length); break; }
    }
    if (/\.git$/.test(text)) text = text.slice(0, -4);
    if (text.indexOf('/tree/') > -1) {
      var cut = text.indexOf('/tree/');
      ref = text.slice(cut + 6);
      text = text.slice(0, cut);
    } else if ((text.match(/@/g) || []).length === 1 && text.charAt(0) !== '@') {
      var at = text.indexOf('@');
      ref = text.slice(at + 1);
      text = text.slice(0, at);
    }
    var parts = text.split('/').filter(Boolean);
    if (parts.length < 2) throw new Error('cannot read an owner/repo out of ' + JSON.stringify(value));
    if (!NAME.test(parts[0]) || !NAME.test(parts[1])) {
      throw new Error(JSON.stringify(value) + ' does not name a GitHub repository');
    }
    if (ref !== null && (!REF.test(ref) || ref.split('/').indexOf('..') > -1)) {
      throw new Error(JSON.stringify(ref) + ' is not a usable git ref');
    }
    return { owner: parts[0], repo: parts[1], ref: ref };
  }

  // package.json -> the same shape `lockfile.read_manifest` produces.
  function readManifest(doc) {
    if (!doc || typeof doc !== 'object' || Array.isArray(doc)) {
      throw new Error('package.json is not an object');
    }
    var direct = {};
    [['dependencies', false], ['devDependencies', true]].forEach(function (pair) {
      var entries = doc[pair[0]];
      if (!entries || typeof entries !== 'object') return;
      Object.keys(entries).forEach(function (name) {
        var spec = String(entries[name]);
        direct[name] = {
          name: name, specifier: spec, version: '', dev: pair[1],
          resolved_name: aliasTarget(spec)
        };
      });
    });
    return { tool: 'manifest', lockfile_version: '', direct: direct, resolved: {} };
  }

  function pkg(dep) { return dep.resolved_name || dep.name; }

  function depRow(dep) {
    return {
      package: dep.name, version: dep.version, specifier: dep.specifier,
      dev: dep.dev, pinned: isPinned(dep.specifier)
    };
  }

  function side(lf) {
    var names = Object.keys(lf.direct);
    var pinned = 0, runtime = 0;
    names.forEach(function (n) {
      if (isPinned(lf.direct[n].specifier)) pinned++;
      if (!lf.direct[n].dev) runtime++;
    });
    return {
      tool: lf.tool, lockfile_version: lf.lockfile_version,
      direct: names.length, runtime: runtime,
      packages: Object.keys(lf.resolved).length, pinned: pinned
    };
  }

  function compare(baseline, subject) {
    var a = baseline.direct, b = subject.direct;
    var aNames = Object.keys(a), bNames = Object.keys(b);

    var addedNames = bNames.filter(function (n) { return !(n in a); }).sort(byCodepoint);
    var removedNames = aNames.filter(function (n) { return !(n in b); }).sort(byCodepoint);
    var shared = aNames.filter(function (n) { return n in b; }).sort(byCodepoint);

    var replaced = [];
    var byBaseRemoved = {}, byBaseAdded = {};
    removedNames.forEach(function (n) { byBaseRemoved[basename(n)] = n; });
    addedNames.forEach(function (n) { byBaseAdded[basename(n)] = n; });
    Object.keys(byBaseRemoved).filter(function (base) {
      return base in byBaseAdded;
    }).sort(byCodepoint).forEach(function (base) {
      var was = byBaseRemoved[base], now = byBaseAdded[base];
      if (was === now) return;
      replaced.push({
        package: base, was: was, now: now,
        was_version: a[was].version, now_version: b[now].version,
        into_scope: scopeOf(now), out_of_scope: scopeOf(was), via_alias: false
      });
    });
    shared.forEach(function (name) {
      if (pkg(a[name]) === pkg(b[name])) return;
      replaced.push({
        package: name, was: pkg(a[name]), now: pkg(b[name]),
        was_version: a[name].version, now_version: b[name].version,
        into_scope: scopeOf(pkg(b[name])), out_of_scope: scopeOf(pkg(a[name])),
        via_alias: true
      });
    });

    var swapped = {}, aliasedKeys = {};
    replaced.forEach(function (r) {
      swapped[r.was] = true; swapped[r.now] = true;
      if (r.via_alias) aliasedKeys[r.package] = true;
    });

    var added = addedNames.filter(function (n) { return !(n in swapped); })
      .map(function (n) { return depRow(b[n]); });
    var removed = removedNames.filter(function (n) { return !(n in swapped); })
      .map(function (n) { return depRow(a[n]); });

    var pinning = [];
    shared.forEach(function (name) {
      if (name in aliasedKeys) return;
      // Version deltas need resolved versions, which a manifest does not have.
      // The section stays empty rather than being filled with specifier text
      // dressed up as a version.
      var oldPinned = isPinned(a[name].specifier), newPinned = isPinned(b[name].specifier);
      if (oldPinned !== newPinned) {
        pinning.push({
          package: name, from: a[name].specifier, to: b[name].specifier,
          direction: newPinned ? 'pinned' : 'unpinned'
        });
      }
    });

    var byKind = { major: 0, minor: 0, patch: 0, downgrade: 0, changed: 0 };
    return {
      baseline: side(baseline), subject: side(subject),
      direct: {
        added: added, removed: removed, replaced: replaced,
        bumped: [], pinning: pinning
      },
      tree: { added: [], removed: [], multi_version: [] },
      totals: {
        direct_baseline: aNames.length, direct_subject: bNames.length,
        added: added.length, removed: removed.length, replaced: replaced.length,
        bumped: 0,
        bumped_major: byKind.major, bumped_minor: byKind.minor,
        bumped_patch: byKind.patch, bumped_downgrade: byKind.downgrade,
        bumped_changed: byKind.changed,
        pinning: pinning.length,
        tree_baseline: 0, tree_subject: 0,
        tree_added: 0, tree_removed: 0, multi_version: 0
      }
    };
  }

  function headline(diff) {
    var t = diff.totals, bits = [];
    if (t.added) bits.push(t.added + ' added');
    if (t.removed) bits.push(t.removed + ' dropped');
    if (t.replaced) bits.push(t.replaced + ' replaced with a scoped republish');
    if (t.bumped) {
      bits.push(t.bumped + ' bumped' + (t.bumped_major ? ' (' + t.bumped_major + ' major)' : ''));
    }
    if (t.bumped_downgrade) bits.push(t.bumped_downgrade + ' downgraded');
    if (t.pinning) bits.push(t.pinning + ' pinning change(s)');
    if (!bits.length) return 'The two sides declare identical direct dependencies.';
    var lead = 'Against the baseline, the subject has ' + bits.join(', ');
    // No lockfile was read, so there is no tree. "0 packages in the resolved
    // tree" would read as a measurement rather than an absence.
    if (!t.tree_baseline && !t.tree_subject) return lead + '.';
    var delta = t.tree_subject - t.tree_baseline;
    return lead + ' — and ' + (delta > 0 ? '+' : '') + delta +
      ' package(s) in the resolved tree (' + t.tree_baseline + ' → ' + t.tree_subject + ').';
  }

  var api = {
    parseRepo: parseRepo, readManifest: readManifest, compare: compare,
    headline: headline, splitIdent: splitIdent, aliasTarget: aliasTarget,
    isPinned: isPinned, basename: basename
  };
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  else global.reconCompare = api;
})(typeof globalThis !== 'undefined' ? globalThis : this);

File: recon/compare.py
"""Compare two repositories' committed lockfiles.

The question this answers is the org's own thesis pointed outward: *a fork is
only worth carrying if the tree is actually healthier than what it forked.*
Given two repos, it says what the second one did to the first — what it added,
dropped, bumped, pinned, and what it replaced with a scoped republish of its
own, which is exactly the `@unabandoned/*` pattern seen from the outside.

Two properties make this cheap and exact where the rest of intake is neither:

* **No resolver.** Both sides already resolved and committed their answers.
* **No registry.** Every fact here comes from two files. Classification needs
  publish dates and is a separate, expensive step; this is not that step, and
  it deliberately reports nothing about health so it cannot imply it.

The sides are `baseline` and `subject`, not `upstream` and `fork`. Nothing here
verifies a fork relationship — the caller asserts the pairing by choosing what
to compare, and calling one side "upstream" would be recon inventing a fact.
"""
from __future__ import annotations

import re

from .lockfile import Dep, Lockfile

MAJOR, MINOR, PATCH = "major", "minor", "patch"
DOWNGRADE, CHANGED = "downgrade", "changed"

_NUM = re.compile(r"^(\d+)\.(\d+)\.(\d+)")


def _parts(version: str) -> tuple[int, int, int] | None:
    m = _NUM.match((version or "").lstrip("v"))
    return (int(m[1]), int(m[2]), int(m[3])) if m else None


def bump_kind(old: str, new: str) -> str:
    """Deliberately simple, and it says `changed` when it cannot tell.

    Prereleases and non-numeric versions are not ranked. Guessing an ordering
    for `2.0.0-rc.1` would produce a confident direction from a comparison that
    does not have one.
    """
    a, b = _parts(old), _parts(new)
    if a is None or b is None or a == b:
        return CHANGED
    if b < a:
        return DOWNGRADE
    if b[0] != a[0]:
        return MAJOR
    if b[1] != a[1]:
        return MINOR
    return PATCH


def basename(name: str) -> str:
    """`@thetechnetwork/composerize-ts` -> `composerize-ts`."""
    return name.split("/", 1)[1] if name.startswith("@") and "/" in name else name


def _scope(name: str) -> str:
    return name.split("/", 1)[0] if name.startswith("@") and "/" in name else ""


def compare(baseline: Lockfile, subject: Lockfile) -> dict:
    """Diff two lockfiles. Pure — no I/O, no registry, no clock."""
    a, b = baseline.direct, subject.direct

    added_names = sorted(set(b) - set(a))
    removed_names = sorted(set(a) - set(b))

    # --- replacements: a package swapped for a scoped republish of itself ----
    # `composerize-ts` leaving while `@thetechnetwork/composerize-ts` arrives is
    # not an unrelated add and drop, it is the same decision the `@unabandoned`
    # scope exists to make. Matched on the unscoped basename, which is the same
    # join `intake` uses for coverage, so the two agree on what "the same
    # package under a different owner" means.
    replaced: list[dict] = []
    by_base_removed = {basename(n): n for n in removed_names}
    by_base_added = {basename(n): n for n in added_names}
    for base in sorted(set(by_base_removed) & set(by_base_added)):
        was, now = by_base_removed[base], by_base_added[base]
        if was == now:
            continue
        replaced.append({
            "package": base,
            "was": was, "now": now,
            "was_version": a[was].version, "now_version": b[now].version,
            "into_scope": _scope(now), "out_of_scope": _scope(was),
            "via_alias": False,
        })
    # An alias keeps the manifest key and swaps the package underneath it —
    # `"buffer": "^5"` becoming `"buffer": "npm:@unabandoned/buffer@^6"`. The key
    # is unchanged on both sides, so nothing above sees it, and it is exactly the
    # adoption move this org makes. Reported as a replacement, which is what it is.
    for name in sorted(set(a) & set(b)):
        if a[name].package == b[name].package:
            continue
        replaced.append({
            "package": name,
            "was": a[name].package, "now": b[name].package,
            "was_version": a[name].version, "now_version": b[name].version,
            "into_scope": _scope(b[name].package),
            "out_of_scope": _scope(a[name].package),
            "via_alias": True,
        })

    swapped = {r["was"] for r in replaced} | {r["now"] for r in replaced}
    aliased_keys = {r["package"] for r in replaced if r.get("via_alias")}

    added = [_dep_row(b[n]) for n in added_names if n not in swapped]
    removed = [_dep_row(a[n]) for n in removed_names if n not in swapped]

    # --- version and pinning deltas on what both sides carry ----------------
    bumped: list[dict] = []
    pinning: list[dict] = []
    for name in sorted(set(a) & set(b)):
        old, new = a[name], b[name]
        # An alias swap is already reported as a replacement. Listing it again
        # as a version bump would compare two different packages' versions and
        # call the result a bump.
        if name in aliased_keys:
            continue
        if old.version != new.version and old.version and new.version:
            bumped.append({
                "package": name,
                "from": old.version, "to": new.version,
                "kind": bump_kind(old.version, new.version),
                "from_specifier": old.specifier, "to_specifier": new.specifier,
                "dev": new.dev,
            })
        if old.pinned != new.pinned:
            pinning.append({
                "package": name,
                "from": old.specifier, "to": new.specifier,
                # "Fix forward, don't pin" is the org's stated position, so the
                # direction is named rather than left for the reader to infer.
                "direction": "pinned" if new.pinned else "unpinned",
            })

    # --- the transitive picture ---------------------------------------------
    tree_added = sorted(set(subject.resolved) - set(baseline.resolved))
    tree_removed = sorted(set(baseline.resolved) - set(subject.resolved))
    multi_version = sorted(
        n for n, vs in subject.resolved.items() if len(vs) > 1
    )

    by_kind = {k: 0 for k in (MAJOR, MINOR, PATCH, DOWNGRADE, CHANGED)}
    for row in bumped:
        by_kind[row["kind"]] += 1

    return {
        "baseline": _side(baseline),
        "subject": _side(subject),
        "direct": {
            "added": added,
            "removed": removed,
            "replaced": replaced,
            "bumped": sorted(bumped, key=lambda r: (
                [MAJOR, DOWNGRADE, MINOR, PATCH, CHANGED].index(r["kind"]), r["package"])),
            "pinning": pinning,
        },
        "tree": {
            "added": tree_added,
            "removed": tree_removed,
            "multi_version": multi_version,
        },
        "totals": {
            "direct_baseline": len(a),
            "direct_subject": len(b),
            "added": len(added),
            "removed": len(removed),
            "replaced": len(replaced),
            "bumped": len(bumped),
            **{f"bumped_{k}": v for k, v in by_kind.items()},
            "pinning": len(pinning),
            "tree_baseline": baseline.total_packages,
            "tree_subject": subject.total_packages,
            "tree_added": len(tree_added),
            "tree_removed": len(tree_removed),
            "multi_version": len(multi_version),
        },
    }


def _side(lf: Lockfile) -> dict:
    return {
        "tool": lf.tool,
        "lockfile_version": lf.lockfile_version,
        "direct": len(lf.direct),
        "runtime": len(lf.runtime),
        "packages": lf.total_packages,
        "pinned": sum(1 for d in lf.direct.values() if d.pinned),
    }


def _dep_row(dep: Dep) -> dict:
    return {
        "package": dep.name, "version": dep.version,
        "specifier": dep.specifier, "dev": dep.dev, "pinned": dep.pinned,
    }


def headline(diff: dict) -> str:
    """One sentence a human can check against the tables below it."""
    t = diff["totals"]
    bits = []
    if t["added"]:
        bits.append(f"{t['added']} added")
    if t["removed"]:
        bits.append(f"{t['removed']} dropped")
    if t["replaced"]:
        bits.append(f"{t['replaced']} replaced with a scoped republish")
    if t["bumped"]:
        majors = t[f"bumped_{MAJOR}"]
        bits.append(
            f"{t['bumped']} bumped" + (f" ({majors} major)" if majors else "")
        )
    if t[f"bumped_{DOWNGRADE}"]:
        bits.append(f"{t[f'bumped_{DOWNGRADE}']} downgraded")
    if t["pinning"]:
        bits.append(f"{t['pinning']} pinning change(s)")
    if not bits:
        return "The two manifests declare identical direct dependencies."
    delta = t["tree_subject"] - t["tree_baseline"]
    sign = "+" if delta > 0 else ""
    return (
        "Against the baseline, the subject has "
        + ", ".join(bits)
        + f" — and {sign}{delta} package(s) in the resolved tree "
          f"({t['tree_baseline']} → {t['tree_subject']})."
    )


# --------------------------------------------------------------------------- #
# The report of record
# --------------------------------------------------------------------------- #
def build_report(baseline_fact, subject_fact, *, baseline_ref: str, subject_ref: str,
                 session, compared_at: str = "", builder_sha: str = "") -> dict:
    """Assemble the comparison, or a report that says why there isn't one."""
    from . import integrity

    sides = {
        "baseline": {"input": baseline_ref, **baseline_fact.provenance()},
        "subject": {"input": subject_ref, **subject_fact.provenance()},
    }
    checks = [_both_sides_read(baseline_fact, subject_fact, sides)]

    report = {
        "schema_version": 1,
        "kind": "compare",
        "meta": {
            "baseline": baseline_ref, "subject": subject_ref,
            "compared_at": compared_at, "builder_sha": builder_sha,
        },
        "sides": sides,
        "coverage": {"fetches": session.summary()},
    }

    if baseline_fact.is_ok and subject_fact.is_ok:
        a, b = baseline_fact.payload, subject_fact.payload
        report["diff"] = compare(a, b)
        report["headline"] = headline(report["diff"])
        checks.append(_same_tool(a, b))
    else:
        # No diff at all, rather than an empty one. An empty diff renders as
        # "these repositories are identical", which is the most confidently
        # wrong thing this report could say.
        report["diff"] = None
        report["headline"] = ""

    report["integrity"] = {
        "status": integrity.worst_status(checks),
        "checks": [c.to_json() for c in sorted(checks, key=lambda c: c.id)],
        "counts": integrity.counts(checks),
    }
    return report


def _both_sides_read(baseline_fact, subject_fact, sides) -> "object":
    from . import integrity
    unread = [
        name for name, fact in (("baseline", baseline_fact), ("subject", subject_fact))
        if not fact.is_ok
    ]
    if unread:
        detail = "; ".join(
            f"{n}: {sides[n].get('detail') or 'unread'}" for n in unread
        )
        return integrity.Check(
            "compare.both-sides-read", "M1", integrity.FAIL,
            "Both lockfiles were read",
            f"{len(unread)} side(s) could not be read — {detail}. No comparison is "
            "reported, because an empty diff reads as 'identical'.",
            {"unread": unread},
        )
    return integrity.Check(
        "compare.both-sides-read", "M1", integrity.PASS,
        "Both lockfiles were read",
        f"baseline {sides['baseline'].get('source', '')} and "
        f"subject {sides['subject'].get('source', '')}",
        {},
    )


def _same_tool(a, b) -> "object":
    from . import integrity
    if a.tool == b.tool:
        return integrity.Check(
            "compare.same-package-manager", "M2", integrity.PASS,
            "Both sides were resolved by the same package manager",
            f"both {a.tool} (lockfile {a.lockfile_version} / {b.lockfile_version})",
            {"tool": a.tool},
        )
    # Direct dependencies stay comparable — they come from the manifest either
    # way. The resolved trees do not: npm and pnpm hoist and dedupe differently,
    # so the package counts measure the tools as much as the projects.
    return integrity.Check(
        "compare.same-package-manager", "M2", integrity.WARN,
        "Both sides were resolved by the same package manager",
        f"baseline uses {a.tool} and subject uses {b.tool} — direct dependencies "
        "remain comparable, but the resolved-tree counts partly measure the "
        "difference between the two resolvers rather than between the projects",
        {"baseline": a.tool, "subject": b.tool},
    )
